### PR TITLE
kvserver: improve redaction for failed consistency checks

### PR DESCRIPTION
**Pull request by tbg**
Release justification: low-risk change to logging
Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_consistency_diff.go [hunk 1]
```diff
@@ -36,46 +35,38 @@ type ReplicaSnapshotDiff struct {
...
-func (rsds ReplicaSnapshotDiffSlice) WriteTo(w io.Writer) (int64, error) {
-	n, err := w.Write([]byte("--- leaseholder\n+++ follower\n"))
-	if err != nil {
-		return 0, err
-	}
+func (rsds ReplicaSnapshotDiffSlice) WriteTo(buf *redact.StringBuilder) {
+	buf.Printf("--- leaseholder\n+++ follower\n")
 	for _, d := range rsds {
-		prefix := "+"
+		prefix := redact.Safe("+")
```
Comment (knz) [inline pkg/kv/kvserver/replica_consistency_diff.go:41]: `SafeString`

File: pkg/kv/kvserver/replica_consistency_diff.go [hunk 2]
```diff
@@ -36,46 +35,38 @@ type ReplicaSnapshotDiff struct {
...
+			prefix = redact.Safe("-")
 		}
 		ts := d.Timestamp
 		const format = `%s%d.%09d,%d %s
 %s    ts:%s
 %s    value:%s
 %s    raw mvcc_key/value: %x %x
 `
-		var prettyTime string
+		var prettyTime redact.SafeValue
```
Comment (knz) [inline pkg/kv/kvserver/replica_consistency_diff.go:52]: ditto

File: pkg/kv/kvserver/replica_consistency_diff.go [hunk 3]
```diff
@@ -36,46 +35,38 @@ type ReplicaSnapshotDiff struct {
...
+			prettyTime = redact.Safe("<zero>")
 		} else {
-			prettyTime = d.Timestamp.GoTime().UTC().String()
+			prettyTime = redact.Safe(d.Timestamp.GoTime().UTC().String())
 		}
 		mvccKey := storage.MVCCKey{Key: d.Key, Timestamp: ts}
-		num, err := fmt.Fprintf(w, format,
-			prefix, ts.WallTime/1e9, ts.WallTime%1e9, ts.Logical, d.Key,
+		buf.Printf(format,
+			prefix, redact.Safe(ts.WallTime/1e9), redact.Safe(ts.WallTime%1e9), redact.Safe(ts.Logical), d.Key,
```
Comment (knz) [inline pkg/kv/kvserver/replica_consistency_diff.go:60]: are you sure these `redact.Safe` calls are necessary? IIRC numeric values are not redactable.

File: pkg/kv/kvserver/stores_server.go [hunk 4]
```diff
@@ -63,7 +64,7 @@ func (is Server) CollectChecksum(
 				// snapshot (if present) intact.
 				if len(req.Checksum) > 0 {
 					log.Errorf(ctx, "consistency check failed on range r%d: expected checksum %x, got %x",
-						req.RangeID, req.Checksum, ccr.Checksum)
+						req.RangeID, redact.Safe(req.Checksum), redact.Safe(ccr.Checksum))
```
Comment (knz) [inline pkg/kv/kvserver/stores_server.go:67]: see question above about numeric values

File: pkg/kv/kvserver/replica_consistency_diff.go [hunk 5]
```diff
@@ -36,46 +35,38 @@ type ReplicaSnapshotDiff struct {
 type ReplicaSnapshotDiffSlice []ReplicaSnapshotDiff
 
 // WriteTo writes a string representation of itself to the given writer.
-func (rsds ReplicaSnapshotDiffSlice) WriteTo(w io.Writer) (int64, error) {
-	n, err := w.Write([]byte("--- leaseholder\n+++ follower\n"))
-	if err != nil {
-		return 0, err
-	}
+func (rsds ReplicaSnapshotDiffSlice) WriteTo(buf *redact.StringBuilder) {
```
Comment (knz) [inline pkg/kv/kvserver/replica_consistency_diff.go:38]: Ugh I don't see the point of implementing `WriteTo`. 

Why don't you make this a standard-looking `SafeFormat` method instead

File: pkg/kv/kvserver/replica_consistency_diff.go [hunk 6]
```diff
@@ -36,46 +35,38 @@ type ReplicaSnapshotDiff struct {
...
 			prefix, storage.EncodeKey(mvccKey), d.Value)
-		if err != nil {
-			return 0, err
-		}
-		n += num
 	}
-	return int64(n), nil
 }
 
 func (rsds ReplicaSnapshotDiffSlice) String() string {
```
Comment (knz) [inline pkg/kv/kvserver/replica_consistency_diff.go:67]: once the thing implements `SafeFormatter`, you can implement this as `{ return redact.StringWithoutMarkers(rsds) }` instead

File: pkg/kv/kvserver/replica_consistency_diff.go (hunk 1 above)
Comment (tbg) [reply pkg/kv/kvserver/replica_consistency_diff.go:41]: Changed - I had that originally but the comment on `SafeString` seemed to indicate using it was somehow "worse" than using `Safe`.

Comment (tbg) [reply pkg/kv/kvserver/replica_consistency_diff.go:41]: (no push yet)

File: pkg/kv/kvserver/stores_server.go (hunk 4 above)
Comment (tbg) [reply pkg/kv/kvserver/stores_server.go:67]: These checksums are byte slices, so these definitely need the `redact.Safe`.

File: pkg/kv/kvserver/replica_consistency_diff.go (hunk 6 above)
Comment (tbg) [reply pkg/kv/kvserver/replica_consistency_diff.go:67]: Done.

**Comment by tbg** [comment]
TFTR, PTAL

**Review by knz — COMMENTED** [review]
> Reviewed 2 of 4 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54108)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/consistency_queue_test.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGn3qpM9hA8GoQx_v9h:-MGn3qpNCT1Pw8MGQvxS:boyzy5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/797b2bd7fc1fb9079b2429aad72c34426f88c105/pkg/kv/kvserver/consistency_queue_test.go#L231)):*
> ```Go
> 
> 	defer log.TestingSetRedactable(true)()
> ```

you sure you want to keep this?
___
*[pkg/kv/kvserver/replica_consistency.go, line 286 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGn40l4-4K0o038vP2l:-MGn40l4-4K0o038vP2m:b-6eqqgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/797b2bd7fc1fb9079b2429aad72c34426f88c105/pkg/kv/kvserver/replica_consistency.go#L286)):*
> ```Go
> 		args.Terminate = append(args.Terminate, results[idxs].Replica)
> 	}
> 	log.Errorf(ctx, "consistency check failed; fetching details and shutting down minority %v", redact.Safe(args.Terminate))
> ```

this one looks suspicious: what guarantees that `args.Terminate` won't be surreptitiously upgraded to a more complex struct with unwanted details?
___
*[pkg/kv/kvserver/replica_consistency_diff.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGmkgURv_qfuumq8Gx9-r1-41:-MGn4B4q1hYJuBq0d370:bmu7pot) ([raw file](https://github.com/cockroachdb/cockroach/blob/f83478005e6585e72fcc6ba2ed40e36bf1db1056/pkg/kv/kvserver/replica_consistency_diff.go#L41)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

(no push yet)
</blockquote></details>

SafeString is fine for literal strings
___
*[pkg/kv/kvserver/replica_consistency_diff.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGmkgURv_qfuumq8Gx9-r1-60:-MGn4_h410DVvfYsbaLQ:bfv6l6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f83478005e6585e72fcc6ba2ed40e36bf1db1056/pkg/kv/kvserver/replica_consistency_diff.go#L60)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

are you sure these `redact.Safe` calls are necessary? IIRC numeric values are not redactable.
</blockquote></details>

ditto: why spell out separate print directives for WallTime / Logical, where ts.String / ts.SafeFormat already do "the right thing"
___
*[pkg/kv/kvserver/replica_consistency_diff.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGn4Dxa-dvu1JAqqKx-:-MGn4Dxa-dvu1JAqqKx0:b-u6b1m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/797b2bd7fc1fb9079b2429aad72c34426f88c105/pkg/kv/kvserver/replica_consistency_diff.go#L56)):*
> ```Go
> 			prettyTime = redact.SafeString("<zero>")
> 		} else {
> 			prettyTime = redact.SafeString(d.Timestamp.GoTime().UTC().String())
> ```

which means that this one may be the problematic one. 

But that said ... I don't get it. hlc.Timestamp has already a SafeFormat and already prints "zero" for zero values. Is `prettyTime` really needed?

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54108)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/kv/kvserver/consistency_queue_test.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGn3qpM9hA8GoQx_v9h:-MGr4NxE9M1D5ut6wjry:b526l7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/797b2bd7fc1fb9079b2429aad72c34426f88c105/pkg/kv/kvserver/consistency_queue_test.go#L231)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you sure you want to keep this?
</blockquote></details>

Yes, why not? This is the best test for actually looking at the diff.
___
*[pkg/kv/kvserver/replica_consistency.go, line 286 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGn40l4-4K0o038vP2l:-MGr4QrdDouTmYL8yZky:buagxv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/797b2bd7fc1fb9079b2429aad72c34426f88c105/pkg/kv/kvserver/replica_consistency.go#L286)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

this one looks suspicious: what guarantees that `args.Terminate` won't be surreptitiously upgraded to a more complex struct with unwanted details?
</blockquote></details>

Fixed with a TODO for you as discussed offline.
___
*[pkg/kv/kvserver/replica_consistency_diff.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGmkgURv_qfuumq8Gx9-r1-60:-MGrBLGN-g-cCXMHxll8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f83478005e6585e72fcc6ba2ed40e36bf1db1056/pkg/kv/kvserver/replica_consistency_diff.go#L60)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto: why spell out separate print directives for WallTime / Logical, where ts.String / ts.SafeFormat already do "the right thing"
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_consistency_diff.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGn4Dxa-dvu1JAqqKx-:-MGrBK4p0iy3n_bpTw7w:bs73i8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/797b2bd7fc1fb9079b2429aad72c34426f88c105/pkg/kv/kvserver/replica_consistency_diff.go#L56)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

which means that this one may be the problematic one. 

But that said ... I don't get it. hlc.Timestamp has already a SafeFormat and already prints "zero" for zero values. Is `prettyTime` really needed?
</blockquote></details>

Fair, done.

**Review by knz — COMMENTED** [review]
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54108)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[docs/generated/redact_safe.md, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGrFROkCfUXOX3y8HJv:-MGrFROkCfUXOX3y8HJw:b-lkmtq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5a0dcdfe1f70b36b99b335e2089295ffe4b54ea/docs/generated/redact_safe.md#L24)):*
> ```Markdown
> pkg/sql/sem/tree/table_ref.go | `ColumnID`
> pkg/storage/enginepb/mvcc3.go | `MVCCStatsDelta`
> pkg/storage/enginepb/mvcc3.go | `*MVCCStats`
> ```

I feel icky about including a struct  in this list. There's no obstacle to adding non-numeric fields into that struct later.

I think best practice would be to add a SafeFormat to that struct instead of turning it into a SafeValue; and the second best is to add a big fat comment in the proto file at the top with "DO NOT ADD ANY NON-NUMERIC FIELDS HERE".


(Side note: my ability to spot this here is exactly why redact_safe.md is generated, and I am glad it exists)

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54108)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[docs/generated/redact_safe.md, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGrFROkCfUXOX3y8HJv:-MGrWYe9-8B44ATsfGo8:b-hxmxcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5a0dcdfe1f70b36b99b335e2089295ffe4b54ea/docs/generated/redact_safe.md#L24)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I feel icky about including a struct  in this list. There's no obstacle to adding non-numeric fields into that struct later.

I think best practice would be to add a SafeFormat to that struct instead of turning it into a SafeValue; and the second best is to add a big fat comment in the proto file at the top with "DO NOT ADD ANY NON-NUMERIC FIELDS HERE".


(Side note: my ability to spot this here is exactly why redact_safe.md is generated, and I am glad it exists)
</blockquote></details>

Wouldn't the best practice be that I don't have to do anything? (https://github.com/cockroachdb/redact/issues/5)?

I'm going to go with the warning for now, as these messages don't even have a custom `String()` method yet and I don't want to make just for redaction if

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54108)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[docs/generated/redact_safe.md, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54108#-MGrFROkCfUXOX3y8HJv:-MGrWYe9-8B44ATsfGo8:b-hxmxcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5a0dcdfe1f70b36b99b335e2089295ffe4b54ea/docs/generated/redact_safe.md#L24)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I feel icky about including a struct  in this list. There's no obstacle to adding non-numeric fields into that struct later.

I think best practice would be to add a SafeFormat to that struct instead of turning it into a SafeValue; and the second best is to add a big fat comment in the proto file at the top with "DO NOT ADD ANY NON-NUMERIC FIELDS HERE".


(Side note: my ability to spot this here is exactly why redact_safe.md is generated, and I am glad it exists)
</blockquote></details>

Wouldn't the best practice be that I don't have to do anything? (https://github.com/cockroachdb/redact/issues/5)?

I'm going to go with the warning for now, as these messages don't even have a custom `String()` method yet and I don't want to make just for redaction if

**Comment by tbg** [comment]
... if we are going to fix up redaction anyway.

**Review by knz — APPROVED** [review]
> Reviewed 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54108)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

**Comment by tbg** [comment]
bors r=knz
Lint failure fixed.
TFTR

**Pull request merged**